cluster.go

```diff
@@ -289,6 +312,74 @@ func parseSlots(slots RedisMessage, defaultAddr string) map[string]group {
 	return groups
 }
 
+// parseShards - map redis shards for each redis nodes/addresses
+// defaultAddr is needed in case the node does not know its own IP
+func parseShards(shards RedisMessage, defaultAddr string) map[string]group {
+	parseNodeEndpoint := func(msg map[string]RedisMessage) string {
+		var ip string
+		switch msg["ip"].string {
+		case "":
+			return defaultAddr
+		case "?":
+			return ""
+		default:
+			ip = msg["ip"].string
+		}
+
+		_, isPort := msg["port"]
+		_, isTlsPort := msg["tls-port"]
+		var port int64
+		switch {
+		case isPort && !isTlsPort:
+			port = msg["port"].integer
+		case !isPort && isTlsPort:
+			port = msg["tls-port"].integer
+		default:
+			// TODO which one to choose if both set?
+			port = msg["port"].integer
+		}
+		return net.JoinHostPort(ip, strconv.FormatInt(port, 10))
+	}
+
+	groups := make(map[string]group, len(shards.values))
+	for _, v := range shards.values {
+		var master string
+		nodes := v.values[3].values
+		// Not specified that 1st entry is always master
```

Comment:
Only ClusterSlots docs mention that master is always first. Although all example messages in ClusterShards keep master as 1st entry I wasn't sure if it should be assumed.

Reply:
Hi @Exca-DK,

We probably need to iterate through the node list and find the first master role.